server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -80,44 +100,91 @@ protected void doExecute(Task task, FieldCapabilitiesRequest request, final Acti
             concreteIndices = indexNameExpressionResolver.concreteIndexNames(clusterState, localIndices);
         }
 
-        checkIndexBlocks(clusterState, concreteIndices);
-
-        final int totalNumRequest = concreteIndices.length + remoteClusterIndices.size();
-        if (totalNumRequest == 0) {
+        if (concreteIndices.length == 0 && remoteClusterIndices.isEmpty()) {
             listener.onResponse(new FieldCapabilitiesResponse(new String[0], Collections.emptyMap()));
             return;
         }
 
-        final List<FieldCapabilitiesIndexResponse> indexResponses = Collections.synchronizedList(new ArrayList<>());
+        checkIndexBlocks(clusterState, concreteIndices);
+
+        final Map<String, FieldCapabilitiesIndexResponse> indexResponses = Collections.synchronizedMap(new HashMap<>());
         final FailureCollector indexFailures = new FailureCollector();
-        final Runnable countDown = createResponseMerger(request, totalNumRequest, indexResponses, indexFailures, listener);
 
+        // If all nodes are on version 7.16 or higher, then we group the shard requests and send a single request per node.
+        // Otherwise, for backwards compatibility we follow the old strategy of sending a separate request per shard.
```

Comment:
We have to take the remote cluster into account. Would it be simpler to make the decision on a per-connection level ? I think it's "ok" to emulate the old model by translating the node request into multiple shard request on older connections.

Reply:
I'm having trouble understanding this part. Could you explain why we need to take the remote cluster into account?